### PR TITLE
Introduce `Client.register_plugin()`

### PR DESCRIPTION
**Pull request by hendrikmakait**
* Follow-up to #8150

This PR introduces `Client.register_plugin()` as the new method to use for registering plugins. Following #8150, it does not allow duck-typed plugins but dispatches based on the inherited plugin class.

It also deprecates `Client.register_scheduler_plugin()` and `Client.register_worker_plugin()` in favor of `Client.register_plugin()`. Within `Client.register_worker_plugin()`, it also adjusts the dispatching logic. With this PR, typing takes precedence, which means that the plugin will get registered based on its inherited plugin class, regardless of the `nanny` keyword and - most notably - if a `SchedulerPlugin` is used, it will be registered as a scheduler plugin. This fixes the migration problem for #8142.

- [x] Tests added / passed
- [x] Passes `pre-commit run --all-files`

### CODE REVIEW

File: distributed/client.py [hunk 1]
```diff
@@ -4902,41 +4978,50 @@ def register_worker_callbacks(self, setup=None):
...
+    def register_worker_plugin(
+        self,
+        plugin: NannyPlugin | WorkerPlugin,
+        name: str | None = None,
+        nanny: bool | None = None,
+    ):
         """
         Registers a lifecycle worker plugin for all current and future workers.
 
+        .. deprecated:: 2023.9.2
```
Comment (hendrikmakait) [inline distributed/client.py:5022]: I'm assuming we do another release in September.

**Comment by hendrikmakait** [comment]
Note that this PR will likely fail if someone uses multiple inheritance, e.g. to subclass both `SchedulerPlugin` and `WorkerPlugin`. I don't think this is a pattern we'd want to encourage, but I've seen it being used at least once.

**Review by fjetter — APPROVED** [review]

**Comment by fjetter** [comment]
changes look good. I could see use cases for multiple inheritance but what I can come up with is a "server plugin". Maybe thi  is something to add later.
The deprecated paths are still available. Let's keep the deprecation warnings in for a bit to see where and if this causes issues

**Comment by hendrikmakait** [comment]
+1 on a `ServerPlugin` that supports a subset of the existing APIs.

> changes look good. I could see use cases for multiple inheritance but what I can come up with is a "server plugin". Maybe thi is something to add later.

+1 on a `ServerPlugin` that supports a subset of the existing APIs.

File: distributed/client.py [hunk 2]
```diff
@@ -3781,10 +3782,11 @@ def upload_file(self, filename, load: bool = True):
 
         async def _():
             results = await asyncio.gather(
-                self.register_scheduler_plugin(
+                self.register_plugin(
                     SchedulerUploadFile(filename, load=load), name=name
                 ),
-                self.register_worker_plugin(UploadFile(filename, load=load), name=name),
+                # FIXME: Make scheduler plugin responsible for (de)registering worker plugin
```
[the remark this answers is not in the document]
Comment (hendrikmakait) [reply distributed/client.py:3788]: #8175

File: distributed/scheduler.py [hunk 3]
```diff
@@ -7483,17 +7491,24 @@ def stop_task_metadata(self, name: str | None = None) -> dict:
         self.remove_plugin(name=plugin.name)
         return {"metadata": plugin.metadata, "state": plugin.state}
 
-    async def register_worker_plugin(self, comm, plugin, name=None):
```
[the remark this answers is not in the document]
Comment (fjetter) [reply distributed/scheduler.py:7486]: no, it can't https://github.com/dask/distributed/pull/8189

**Pull request merged**